src/edu/washington/escience/myria/operator/CSVFileScanFragment.java

```diff
@@ -181,142 +187,184 @@ protected TupleBatch fetchNextReady() throws IOException, DbException {
+         * If the previous character is not a new line, we discard the line.
+         */
+        if (lineNumber - 1 == 0 && partitionStartByteRange != 0) {
+          char currentChar = (char) partitionInputStream.read();
+          if (currentChar != '\n' && currentChar != '\r') {
+            discardedRecord = true;
+            /*
+             * This handles closes the stream in corner cases in which a worker might only have a fragment of a tuple as
+             * its entire partition. Since the server makes sure to provide workers a partition with a minimum size,
+             * this case could only happen with very large rows.
+             */
+            if (onLastRow) {
+              parser.close();
               break;
             }
-            dataChar = startStreamReader.read();
+          } else {
+            /*
+             * This handles the case in which the splitting occurs in the middle of \r\n.
+             */
+            currentChar = (char) partitionInputStream.read();
+            if (currentChar == '\n') {
+              discardedRecord = true;
+            } else {
+              partitionInputStream =
+                  source.getInputStream(adjustedStartByteRange, adjustedEndByteRange);
+            }
           }
-          startStreamReader.close();
         }
-      }
-      // This is a partial fragment, columns are missing and we need to read more bytes
-      if (record.size() < schema.numColumns()) {
-        if (lineNumber - 1 != 0 && !isLastWorker) {
-          onLastRow = true;
-          long bytePositionAtBeginningOfRecord = record.getCharacterPosition();
-          if (!fixingStartByte) {
-            fixingStartByte = true;
-            partitionStartByteRange += bytePositionAtBeginningOfRecord;
+
+        /*
+         * Here, if we are on the last row, we make sure to read the entire row until we either hit a new line or until
+         * we have read the entire file (again, this is for the case where a single worker might be reading a single
+         * large row that was split among other workers)
+         */
+        if (record.size() == schema.numColumns() && onLastRow && !isLastWorker) {
+          long movingEndByte = adjustedEndByteRange;
+          long characterPositionAtBeginningOfRecord = record.getCharacterPosition();
+          boolean newLineFound = false;
+          while (!newLineFound) {
+            movingEndByte += byteOverlap;
+            InputStream trailingEndInputStream =
+                source.getInputStream(
+                    adjustedEndByteRange + 1, Math.min(movingEndByte, maxByteRange));
+            int dataChar = trailingEndInputStream.read();
+            while (dataChar != -1) {
+              char currentChar = (char) dataChar;
+              if (currentChar == '\n'
+                  || currentChar == '\r'
+                  || Math.min(movingEndByte, maxByteRange) == maxByteRange) {
```

Comment:
if `movingEndByte >= maxByteRange` then there's no reason to retrieve a new `trailingEndInputStream` and scan it for newlines, right? We should just move along to constructing the `partitionInputStream` below.